collector/systemd.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nosystemd
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type systemdCollector struct {
+	conn           *dbus.Conn
+	unitStateCount *prometheus.GaugeVec
+	unitFailed     *prometheus.GaugeVec
+	unitFailedMap  map[string]float64
+}
+
+func init() {
+	Factories["systemd"] = NewSystemdCollector
+}
+
+// Takes a prometheus registry and returns a new Collector exposing
+// systemd statistics.
+func NewSystemdCollector() (Collector, error) {
+	conn, err := dbus.New()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get dbus connection: %s", err)
+	}
+	return &systemdCollector{
+		conn: conn,
+		unitStateCount: prometheus.NewGaugeVec(
```

Comment:
You should use ConstMetrics like filesystem collector, rather than sharing metrics across scrapes. This handles deleted services and concurrent scrapes better